// vscode-extension-telemetry-wrapper/lib/Session.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const uuid = require("uuid");
const ExitCode_1 = require("./ExitCode");
class Session {
    constructor(action) {
        this.extraProperties = {};
        this.extraMeasures = {};
        this.id = uuid.v4();
        this.action = action;
        this.startAt = new Date();
    }
    /**
     * @deprecated
     */
    getCustomEvent() {
        const ret = {};
        const extraPropertiesObject = Object.assign({}, ...Object.keys(this.extraProperties).map((k) => ({ [`extra.${k}`]: this.extraProperties[k] })));
        const extraMeasuresObject = Object.assign({}, ...Object.keys(this.extraMeasures).map((k) => ({ [`extra.${k}`]: this.extraMeasures[k] })));
        ret.properties = Object.assign({}, extraPropertiesObject, { sessionId: this.id, action: this.action, startAt: this.startAt });
        ret.measures = Object.assign({}, extraMeasuresObject, { duration: (this.stopAt || new Date()).getTime() - this.startAt.getTime() });
        return ret;
    }
    /**
     * @deprecated
     */
    end() {
        this.stopAt = new Date();
        this.exitCode = this.exitCode || ExitCode_1.ExitCode.SUCCESS;
    }
}
exports.Session = Session;
